File: src/aci_cache/tracker.py
"""Sliding-window write rate tracker for aci-cache."""

from __future__ import annotations

import threading
import time
from typing import List


class WriteRateTracker:
    """Tracks write timestamps in a sliding window to calculate writes/sec.

    Thread-safe: the main thread appends via ``record()`` and the controller
    thread reads via ``get_rate()``.
    """

    def __init__(self, window: float = 5.0) -> None:
        if window <= 0:
            raise ValueError(f"window must be positive, got {window}")
        self._window = window
        self._timestamps: List[float] = []
        self._lock = threading.Lock()

    def record(self) -> None:
        """Record a write at the current time."""
        now = time.time()
        with self._lock:
            self._timestamps.append(now)

    def get_rate(self) -> float:
        """Return the current write rate (writes per second).

        Trims timestamps older than the sliding window before computing.
        """
        now = time.time()
        cutoff = now - self._window
        with self._lock:
            # Trim expired entries from the front
            while self._timestamps and self._timestamps[0] < cutoff:
                self._timestamps.pop(0)
            count = len(self._timestamps)
        return count / self._window

    def get_count(self) -> int:
        """Return the number of writes currently in the window."""
        now = time.time()
        cutoff = now - self._window
        with self._lock:
            while self._timestamps and self._timestamps[0] < cutoff:
                self._timestamps.pop(0)
            return len(self._timestamps)

File: node/src/tracker.ts
/**
 * Sliding-window write rate tracker for aci-cache.
 *
 * No locking needed — Node.js is single-threaded.
 */
export class WriteRateTracker {
  private readonly window: number;
  private timestamps: number[] = [];

  constructor(window = 5) {
    if (window <= 0) {
      throw new Error(`window must be positive, got ${window}`);
    }
    this.window = window;
  }

  /** Record a write at the current time. */
  record(): void {
    this.timestamps.push(Date.now() / 1000);
  }

  /** Return the current write rate (writes per second). */
  getRate(): number {
    this.trim();
    return this.timestamps.length / this.window;
  }

  /** Return the number of writes currently in the window. */
  getCount(): number {
    this.trim();
    return this.timestamps.length;
  }

  /** Remove timestamps older than the sliding window. */
  private trim(): void {
    const cutoff = Date.now() / 1000 - this.window;
    while (this.timestamps.length > 0 && this.timestamps[0] < cutoff) {
      this.timestamps.shift();
    }
  }
}
